File: packages/engine-freecut/src/executor/ffprobe.ts
// packages/engine-freecut/src/executor/ffprobe.ts
import { execFile } from "child_process";
import { promisify } from "util";
const pexec = promisify(execFile);

export async function probeDuration(filePath: string): Promise<number> {
  const { stdout } = await pexec("ffprobe", [
    "-v", "error",
    "-show_entries", "format=duration",
    "-of", "default=noprint_wrappers=1:nokey=1",
    filePath,
  ]);
  const dur = parseFloat(stdout.trim());
  if (!isFinite(dur)) throw new Error(`ffprobe failed for ${filePath}`);
  return dur;
}

File: monet/engines/freecut/executor/ffprobe.py
# monet/engines/freecut/executor/ffprobe.py
from __future__ import annotations
import asyncio
import json


async def probe_duration(file_path: str) -> float:
    proc = await asyncio.create_subprocess_exec(
        "ffprobe", "-v", "error",
        "-show_entries", "format=duration",
        "-of", "default=noprint_wrappers=1:nokey=1",
        file_path,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    out, err = await proc.communicate()
    if proc.returncode != 0:
        raise RuntimeError(f"ffprobe failed for {file_path}: {err.decode()}")
    return float(out.decode().strip())


async def probe_streams(file_path: str) -> dict:
    proc = await asyncio.create_subprocess_exec(
        "ffprobe", "-v", "error",
        "-show_streams", "-show_format",
        "-of", "json", file_path,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    out, err = await proc.communicate()
    if proc.returncode != 0:
        raise RuntimeError(f"ffprobe failed: {err.decode()}")
    return json.loads(out.decode())
